nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/inotify/GetHDFSEvents.java

```diff
@@ -0,0 +1,285 @@
+        props.add(POLL_DURATION);
+        props.add(HDFS_PATH_TO_WATCH);
+        props.add(RECURSE_SUBDIRECTORIES);
+        props.add(EVENT_TYPES);
+        return Collections.unmodifiableList(props);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final StateManager stateManager = context.getStateManager();
+
+        try {
+            StateMap state = stateManager.getState(Scope.CLUSTER);
+            String txIdAsString = state.get(LAST_TX_ID);
+
+            if (txIdAsString != null && !"".equals(txIdAsString)) {
+                lastTxId = Long.parseLong(txIdAsString);
+            }
+        } catch (IOException e) {
+            getLogger().error("Unable to retrieve last transaction ID. Must retrieve last processed transaction ID before processing can occur.", e);
+            context.yield();
+            return;
+        }
+
+        try {
+            final TimeUnit pollDurationTimeUnit = TimeUnit.MICROSECONDS;
+            final long pollDuration = context.getProperty(POLL_DURATION).asTimePeriod(pollDurationTimeUnit);
+            final DFSInotifyEventInputStream eventStream = lastTxId == -1L ? getHdfsAdmin().getInotifyEventStream() : getHdfsAdmin().getInotifyEventStream(lastTxId);
+            final EventBatch eventBatch = getEventBatch(eventStream, pollDuration, pollDurationTimeUnit);
+
+            if (eventBatch != null && eventBatch.getEvents() != null) {
+                if (eventBatch.getEvents().length > 0) {
+                    List<FlowFile> flowFiles = new ArrayList<>(eventBatch.getEvents().length);
+                    for (Event e : eventBatch.getEvents()) {
+                        if (toProcessEvent(context, e)) {
+                            getLogger().debug("Creating flow file for event.");
+                            FlowFile flowFile = session.create();
+                            flowFile = session.putAttribute(flowFile, EventAttributes.MIME_TYPE, "application/json");
+                            flowFile = session.putAttribute(flowFile, EventAttributes.EVENT_TYPE, e.getEventType().name());
+                            flowFile = session.putAttribute(flowFile, EventAttributes.EVENT_PATH, getPath(e));
+                            flowFile = session.write(flowFile, new OutputStreamCallback() {
+                                @Override
+                                public void process(OutputStream out) throws IOException {
+                                    out.write(OBJECT_MAPPER.writeValueAsBytes(e));
+                                }
+                            });
+                            flowFiles.add(flowFile);
+                        }
+                    }
+
+                    getLogger().debug("Transferring all event flow files.");
+                    session.transfer(flowFiles, REL_SUCCESS);
```

Comment:
It'd be nice to add the creation of a provenance event with something like `session.getProvenanceReporter().receive(...)` with some details.